knotx-core/knotx-template-engine/src/main/java/com/cognifide/knotx/engine/service/ServiceEngine.java

```diff
@@ -44,14 +47,20 @@ public ServiceEngine(Vertx vertx, TemplateEngineConfiguration serviceConfigurati
         this.configuration = serviceConfiguration;
     }
 
-    public Observable<Map<String, Object>> doServiceCall(ServiceEntry serviceEntry, MultiMap headers, HttpMethod httpMethod) {
+    public Observable<Map<String, Object>> doServiceCall(ServiceEntry serviceEntry, MultiMap headers, HttpMethod httpMethod, MultiMap formAttributes) {
```

Comment:
Well... I'd refactor this method a little bit, and use the approach from vert.x itself. THe observable created here suppose to move to kind of helper class, let's call it KnotxRxHelper in which we can but a metod for doing request that returns observable, in the form as below (it's just a draft - please double check what I have written )
```
    private Observable<HttpClientResponse> request(HttpClient client, HttpMethod method, int port, String domain, String uri, Action1<HttpClientRequest> requestBuilder) {
        return Observable.create(subscriber -> {
            HttpClientRequest req = client.request(method, port, domain, uri);
            requestBuilder.call(req);
            Observable<HttpClientResponse> resp = req.toObservable();
            resp.subscribe(subscriber);
            req.end();
        });
    }
```
Last parameter is the function call so we can use it to actually compose the request appropriately. E.g. we can invoke our helper:
```
        Observable<HttpClientResponse> serviceResponse = KnotxRxHelper.request(vertx.createHttpClient(), httpMethod, serviceEntry.getPort(), serviceEntry.getDomain(), serviceEntry.getServiceUri(),
                req -> buildRequestBody(req, headers, formAttributes));
```
While buildRequestBody() method is:
```
    private void buildRequestBody(HttpClientRequest request, MultiMap headers, MultiMap formAttributes) {
        request.headers().addAll(headers);
        if (!formAttributes.isEmpty()) {
            Buffer buffer = createFormPostBody(formAttributes);
            request.headers().set("content-length", String.valueOf(buffer.length()));
            request.headers().set("content-type", "application/x-www-form-urlencoded");
            request.write(buffer);
        }
```